modules/renter/downloaddestination.go

```diff
@@ -37,20 +39,56 @@ type downloadDestination interface {
 
 // downloadDestinationBuffer writes logical chunk data to an in-memory buffer.
 // This buffer is primarily used when performing repairs on uploads.
-type downloadDestinationBuffer []byte
+type downloadDestinationBuffer [][]byte
+
+// NewDownloadDestinationBuffer allocates the necessary number of shards for
+// the downloadDestinationBuffer and returns the new buffer.
+func NewDownloadDestinationBuffer(length uint64) downloadDestinationBuffer {
+	// Round length up to next multiple of SectorSize.
+	if length%pieceSize != 0 {
+		length += pieceSize - length%pieceSize
+	}
+	buf := make([][]byte, 0, length/pieceSize)
+	for length > 0 {
+		buf = append(buf, make([]byte, pieceSize))
+		length -= pieceSize
+	}
+	return buf
+}
 
 // Close implements Close for the downloadDestination interface.
 func (dw downloadDestinationBuffer) Close() error {
 	return nil
 }
 
+// ReadFrom reads data from a io.Reader until the buffer is full.
+func (dw downloadDestinationBuffer) ReadFrom(r io.Reader) (int64, error) {
+	var n int64
+	for len(dw) > 0 {
+		read, err := r.Read(dw[0])
+		if err != nil {
+			return 0, err
+		}
+		dw = dw[1:]
+		n += int64(read)
+	}
+	return n, nil
+}
+
 // WriteAt writes the provided data to the downloadDestinationBuffer.
 func (dw downloadDestinationBuffer) WriteAt(data []byte, offset int64) (int, error) {
-	if len(data)+int(offset) > len(dw) || offset < 0 {
+	if uint64(len(data)+int(offset)) > uint64(len(dw))*modules.SectorSize || offset < 0 {
```

Comment:
is there any reason that you have to wrap these in a `uint64`?

Reply:
Yeah the right side of the equation is a `uint64`, `len` is a `ìnt` and `offset` is a int64.